File: src/utils/download-services.js
"use strict";
const fs = require('fs');
const pathLib = require('path');
const axios = require("axios");

async function downloadStreamFile(ctx, url, path, headers = undefined) {
  // Create the directory if it doesn't exist
  const dirPath = pathLib.dirname(path);
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }

  let isSuccessful = false; // Variable to track download status

  const response = await axios({
    method: 'GET',
    url: url,
    responseType: 'stream',
    headers: headers
  });

  response.data.pipe(fs.createWriteStream(path));

  return new Promise((resolve, reject) => {
    response.data.on('end', () => {
      ctx.log(`downloadFile succeeded`);
      isSuccessful = true;
      resolve();
    });

    response.data.on('error', (err) => {
      reject(err);
    });
  }).then(() => {
    return isSuccessful; // Return the download status outside the callbacks
  });
}

module.exports = {
    downloadStreamFile,
};

File: src/utils/download_services.py
import os
import requests
from pathlib import Path

def download_stream_file(ctx, url, path, headers=None):
    # Create the directory if it doesn't exist
    dir_path = Path(path).parent
    os.makedirs(dir_path, exist_ok=True)

    is_successful = False  # Variable to track download status

    response = requests.get(url, headers=headers, stream=True)

    with open(path, 'wb') as file:
        for chunk in response.iter_content(chunk_size=8192):
            file.write(chunk)

    if response.status_code == 200:
        ctx.log("downloadFile succeeded")
        is_successful = True

    return is_successful

